client/js/contextMenuFactory.js

```diff
@@ -0,0 +1,204 @@
+"use strict";
+const $ = require("jquery");
+const socket = require("./socket");
+const utils = require("./utils");
+const JoinChannel = require("./join-channel");
+const ContextMenu = require("./contextMenu");
+const contextMenuActions = [];
+const contextMenuItems = [];
+
+module.exports = {
+	addContextMenuItem,
+	createContextMenu,
+};
+
+addDefaultItems();
+
+/**
+ *
+ * @param opts
+ * @param {function(Object)} [opts.check] - Function to check whether item should show on the context menu, called with the target jquery element, shows if return is truthy
+ * @param {string|function(Object)} opts.className - class name for the menu item, should be prefixed for non-default menu items (if function, called with jquery element, and uses return value)
+ * @param {string|function(Object)} opts.data - data that will be sent to the callback function (if function, called with jquery element, and uses return value)
+ * @param {string|function(Object)} opts.display - text to display on the menu item (if function, called with jquery element, and uses return value)
+ * @param {function(Object)} opts.callback - Function to call when the context menu item is clicked, called with the data requested in opts.data
+ */
+function addContextMenuItem(opts) {
+	opts.check = opts.check || (() => true);
+	opts.actionId = contextMenuActions.push(opts.callback) - 1;
+	contextMenuItems.push(opts);
+}
+
+function addContextDivider(opts) {
+	opts.check = opts.check || (() => true);
+	opts.divider = true;
+	contextMenuItems.push(opts);
+}
+
+function createContextMenu(that, event) {
+	return new ContextMenu(contextMenuItems, contextMenuActions, that, event);
+}
+
+function addWhoisItem() {
+	function whois(itemData) {
+		const chan = utils.findCurrentNetworkChan(itemData);
+
+		if (chan.length) {
+			chan.click();
+		}
+
+		socket.emit("input", {
+			target: $("#chat").data("id"),
+			text: "/whois " + itemData,
+		});
+
+		$(`.channel.active .userlist .user[data-name="${itemData}"]`).trigger("click");
+	}
+
+	addContextMenuItem({
+		check: (target) => target.hasClass("user"),
```

Comment:
Alright, I'm looking only at the exposed API first, because that's what really important.

One question: this uses a single context menu to which we add everything, and each item has its own `check` that we use at render time to decide if it needs to be displayed or hidden, depending on what the target is, correct?

I think it would make for a nicer API if we exposed multiple context menus depending on the context, and when writing a package, one would add items to their expected locations (`channelListMenu.addItem(...)`, `userMenu.addItem(...)`, etc.).
That would also allow for more "specialized" context menus. Totally random example, but imagine context menus for users that would contain specific styling and organization to include avatars or whatnot (I know this is not possible, I said it was random lol). Such context menu could extend the main context menu, or not, but the API would be the same, just a different container.
I'm sure there are ways to make this work with the current solution, just weighing in options to pick what seems the most sensible.

I totally understand if you don't want to bother with that since you already did the work, and also because this PR is merely a modularization of what we already have, but since this is a rather different API shift, I just wanted to mention that before we ship. Whatever the answer is fine as long as we considered our options :)

Reply:
Yeah, so I had thought about that, but I decided not to go for it because currently we don't define different kinds of context menu, so I think it would probably require quiet a bit more refactoring.

I do, however, think this is a good thing. How would you feel about leaving that till a future PR but not adding this functionality to the plugin API until it is done. I don't want to put too much in this because it's quite a refactor as it is, but it doesn't change anything underlying. I'm perfectly happy with doing what you suggest before we add it to the plugin API, but I think it's probably a separate PR. Does that make sense?

Reply:
> How would you feel about leaving that till a future PR but not adding this functionality to the plugin API until it is done.

Completely fine. I totally realize that what I'm describing here is a complete shift from what we currently have compared to your PR which is "just" exposing the current capabilities via an API. This would require changing _a lot_ of things underneath, and it's a PR on its own (even if we were not thinking of packages, we'd possibly consider this on its own).

Reason why I brought that up now is because it'll be a rather different API and I wanted to make sure we're on the same page. Our package APIs are probably going to be unstable for a while anyway, so that's fine.